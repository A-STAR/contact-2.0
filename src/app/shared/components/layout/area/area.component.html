<ng-content></ng-content>

<div
  *ngFor="let i of gutters"
  [ngClass]="getGutterClass()"
  [ngStyle]="getGutterStyle(i)"
  (mousedown)="onMouseDown(i, $event)"
></div>
