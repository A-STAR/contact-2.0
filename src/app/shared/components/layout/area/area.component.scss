:host {
  display: flex;
  flex: 1;
  height: 100%;
  width: 100%;
}

.gutter {
  align-items: center;
  display: flex;
  flex: 0 0 20px;
  justify-content: center;

  .handle {
    background-color: #ddd;
    cursor: ew-resize;
    flex: 0 0 auto;
    height: 48px;
    width: 6px;

    &.horizontal {
      cursor: ns-resize;
      height: 6px;
      width: 48px;
    }
  }
}
