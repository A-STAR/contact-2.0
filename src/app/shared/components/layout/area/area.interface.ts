export enum IAreaLayout {
  ROW    = 'row',
  COLUMN = 'column',
}

export interface IDragData {
  i: number;
  lSize: number;
  rSize: number;
  start: number;
}
