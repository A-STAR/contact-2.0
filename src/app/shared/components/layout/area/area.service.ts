import { Injectable } from '@angular/core';

import { PersistenceService } from '@app/core/persistence/persistence.service';

@Injectable()
export class AreaService {
  constructor(
    private persistenceService: PersistenceService,
  ) {}

  clearState(persistenceKey: string): void {
    this.persistenceService.remove(persistenceKey);
  }

  getState(persistenceKey: string, id: string): number {
    if (id) {
      const state = this.persistenceService.get(persistenceKey);
      return state ? state[id] : null;
    }
    return null;
  }

  saveState(persistenceKey: string, id: string, size: number): void {
    const state = this.persistenceService.get(persistenceKey) || {};
    this.persistenceService.set(persistenceKey, { ...state, [id]: size });
  }
}
