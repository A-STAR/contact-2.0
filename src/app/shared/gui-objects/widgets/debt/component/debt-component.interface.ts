export interface IDebtComponent {
  id: number;
  typeCode: number;
  sum: number;
  currencyId: number;
}
