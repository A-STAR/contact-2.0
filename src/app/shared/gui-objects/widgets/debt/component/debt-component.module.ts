import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { DebtComponentGridModule } from './grid/debt-component-grid.module';

import { DebtComponentService } from './debt-component.service';

@NgModule({
  imports: [
    DebtComponentGridModule,
    CommonModule,
  ],
  exports: [
    DebtComponentGridModule,
  ],
  providers: [
    DebtComponentService,
  ]
})
export class DebtComponentModule { }
