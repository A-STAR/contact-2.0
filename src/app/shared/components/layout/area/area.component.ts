import {
  AfterViewInit,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ContentChildren,
  ElementRef,
  HostBinding,
  Input,
  QueryList,
  Renderer2,
} from '@angular/core';

import { IAreaLayout, IDragData } from './area.interface';

import { range } from '@app/core/utils';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'app-area',
  styleUrls: [ './area.component.scss' ],
  templateUrl: './area.component.html',
})
export class AreaComponent implements AfterViewInit {
  @ContentChildren(AreaComponent, { descendants: false }) _children: QueryList<AreaComponent>;

  @HostBinding('style.flex-direction')
  @Input()
  layout = IAreaLayout.COLUMN;

  @Input()
  persistenceKey: string;

  private mouseMoveListener: () => void;
  private mouseUpListener: () => void;

  private dragData: IDragData;

  get children(): AreaComponent[] {
    return this._children.filter(c => c !== this);
  }

  set order(order: number) {
    this.renderer.setStyle(this.elRef.nativeElement, 'order', order);
  }

  get gutters(): number[] {
    const n = this.children.length;
    return n > 1 ? range(0, n - 2) : [];
  }

  constructor(
    private cdRef: ChangeDetectorRef,
    private elRef: ElementRef,
    private renderer: Renderer2,
  ) {}

  ngAfterViewInit(): void {
    this.children.forEach((c, i) => c.order = i);
  }

  getSize(direction: IAreaLayout): number {
    const r = this.elRef.nativeElement.getBoundingClientRect();
    return direction === IAreaLayout.ROW
      ? r.width
      : r.height;
  }

  setSize(size: number): void {
    this.renderer.setStyle(this.elRef.nativeElement, 'flex', `0 0 ${size}px`);
  }

  getGutterStyle(i: number): Partial<CSSStyleDeclaration> {
    return {
      order: `${i}`,
    };
  }

  onMouseDown(i: number, event: MouseEvent): void {
    const start = this.layout === IAreaLayout.ROW
      ? event.clientX
      : event.clientY;
    this.dragData = {
      start,
      i,
      a: this.children[i].getSize(this.layout),
      b: this.children[i + 1].getSize(this.layout),
    };
    this.mouseMoveListener = this.renderer.listen(this.elRef.nativeElement, 'mousemove', this.onMouseMove.bind(this));
    this.mouseUpListener = this.renderer.listen(this.elRef.nativeElement, 'mouseup', this.onMouseUp.bind(this));
  }

  private onMouseMove(event: MouseEvent): void {
    this.onDrag(event);
  }

  private onMouseUp(event: MouseEvent): void {
    this.onDrag(event);
    this.mouseMoveListener();
    this.mouseUpListener();
  }

  private onDrag(event: MouseEvent): void {
    const { i, start, a, b } = this.dragData;

    const size = this.layout === IAreaLayout.ROW
      ? event.clientX - start
      : event.clientY - start;

    this.children[i].setSize(a + size);
    this.children[i + 1].setSize(b - size);
  }
}
