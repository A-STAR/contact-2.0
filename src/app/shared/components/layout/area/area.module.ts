import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { AreaService } from './area.service';

import { AreaComponent } from './area.component';

@NgModule({
  declarations: [
    AreaComponent,
  ],
  exports: [
    AreaComponent,
  ],
  imports: [
    CommonModule,
  ],
  providers: [
    AreaService,
  ]
})
export class AreaModule {}
