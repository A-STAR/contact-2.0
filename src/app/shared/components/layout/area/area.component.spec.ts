import { ChangeDetectionStrategy, Component } from '@angular/core';
import { async as Async, TestBed } from '@angular/core/testing';
import { EmptyObservable } from 'rxjs/observable/EmptyObservable';

import { AreaService } from './area.service';
import { SettingsService } from '@app/core/settings/settings.service';

import { AreaComponent } from './area.component';

class AreaServiceMock {
  getState(): number {
    return null;
  }
  saveState(): void {
  }
}

class SettingsServiceMock {
  onClear$ = new EmptyObservable();
}

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <app-area layout="column">
      <app-area initialSize="1">
        1
      </app-area>
      <app-area initialSize="2">
        2
      </app-area>
    </app-area>
  `,
})
class AreaColumnContainerComponent {}

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <app-area layout="row">
      <app-area initialSize="1">
        1
      </app-area>
      <app-area initialSize="2">
        2
      </app-area>
    </app-area>
  `,
})
class AreaRowContainerComponent {}

describe('AreaComponent', () => {
  beforeEach(Async(() => {
    TestBed
      .configureTestingModule({
        declarations: [
          AreaComponent,
          AreaColumnContainerComponent,
          AreaRowContainerComponent,
        ],
        providers: [
          {
            provide: AreaService,
            useClass: AreaServiceMock,
          },
          {
            provide: SettingsService,
            useClass: SettingsServiceMock,
          }
        ],
      })
      .compileComponents();
  }));

  it('should render', () => {
    const fixture = TestBed.createComponent(AreaComponent);
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render column layout', () => {
    const fixture = TestBed.createComponent(AreaColumnContainerComponent);
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render row layout', () => {
    const fixture = TestBed.createComponent(AreaRowContainerComponent);
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
