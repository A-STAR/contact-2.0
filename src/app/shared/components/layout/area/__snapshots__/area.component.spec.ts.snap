// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreaComponent should render 1`] = `
<div
  id="root0"
  ng-version="5.2.4"
>
  


  
  

</div>
`;

exports[`AreaComponent should render column layout 1`] = `
<div
  id="root1"
  ng-version="5.2.4"
>
  
    
  <app-area
    layout="column"
    ng-reflect-layout="column"
  >
    
      
    <app-area
      initialsize="1"
      ng-reflect-initial-size="1"
    >
      
        1
      
      


      
      

    </app-area>
    
      
    <app-area
      initialsize="2"
      ng-reflect-initial-size="2"
    >
      
        2
      
      


      
      

    </app-area>
    
    
    


    
    <div
      class="gutter horizontal"
      ng-reflect-ng-class="[object Object]"
      ng-reflect-ng-style="[object Object]"
    />
    

  </app-area>
  
  
</div>
`;

exports[`AreaComponent should render row layout 1`] = `
<div
  id="root2"
  ng-version="5.2.4"
>
  
    
  <app-area
    layout="row"
    ng-reflect-layout="row"
  >
    
      
    <app-area
      initialsize="1"
      ng-reflect-initial-size="1"
    >
      
        1
      
      


      
      

    </app-area>
    
      
    <app-area
      initialsize="2"
      ng-reflect-initial-size="2"
    >
      
        2
      
      


      
      

    </app-area>
    
    
    


    
    <div
      class="gutter vertical"
      ng-reflect-ng-class="[object Object]"
      ng-reflect-ng-style="[object Object]"
    />
    

  </app-area>
  
  
</div>
`;
